File: src/src/libgql/parsers/schema/nodes/union.cpp
#include "./union.hpp"

#include <map>
#include <memory>
#include <ranges>
#include <string>
#include <utility>

#include "../../file/server/ast.hpp"
#include "../../file/shared/ast.hpp"
#include "../server_ast.hpp"
#include "../type_registry.hpp"

namespace parsers::schema::nodes {
std::shared_ptr<ast::Union> parseUnion(
    const file::server::ast::UnionDefinitionNode &node,
    const TypeRegistry &registry) {
    const auto &obj = registry.unions.at(node.name.name);
    obj->items =
        node.values |
        std::views::transform(
            [&registry](const file::shared::ast::NameNode &nNode)
                -> std::pair<std::string, std::shared_ptr<ast::ObjectType>> {
                return { nNode.name, registry.getObject(nNode.name) };
            }) |
        std::ranges::to<std::map>();
    return obj;
};
};  // namespace parsers::schema::nodes

File: rust/libgql/src/parsers/schema/server/union.rs
use crate::parsers::{file, schema::server::errors};

use super::type_registry::HashMapTypeRegistry;

pub fn parse_definition<'buffer>(
    node: &file::server::ast::UnionDefinitionNode<'buffer>,
    registry: &mut HashMapTypeRegistry,
) -> Result<(), errors::Error<'buffer>> {
    let obj = registry.unions.get_mut(node.name.name).unwrap();
    for item in node.values.iter() {
        if let None = registry.objects.get(item.name) {
            return Err(errors::Error::UnknownObject(item.clone()));
        };
        obj.items.insert(item.name.to_string());
    }
    return Ok(());
}
